### ISSUE TITLE
Intance model properties are undefined

### ISSUE THREAD

**Issue by gideaoms** [RobinBuschmann/sequelize-typescript#778]
**Versions**

- sequelize: 5.21.7
- sequelize-typescript:  1.1.0
- typescript: 3.8.3

**I'm submitting a ...** 

[ x ] bug report
[ ] feature request

**Actual behavior:**
I execute a `findOne` function to get a model instance, but the properties are `undefined`

**Expected behavior:**
I would like to get the properties values

**Steps to reproduce:**
```js
const city = await CityModel.findOne({ where: { name: 'example' } })
console.log(city.name);
```

**Related code:**
If I execute the `city.getDataValue('name')` I see the right value, but simply running `city.name` I get `undefined`.

If I execute `console.log(city)` all works well:
```js
CityModel {
    dataValues: {
        id: 'fb73e40a-9055-42b8-9a21-d8368ef13222',
        name: 'Example',
        createdAt: 2020-04-27,
        updatedAt: 2020-04-27
    },
    _previousDataValues: {
        id: 'fb73e40a-9055-42b8-9a21-d8368ef13222',
        name: 'Example',
        createdAt: 2020-04-27,
        updatedAt: 2020-04-27
    },
    ...
},
```

My model:
```ts
import {
  Model,
  Table,
  DataType,
  Column,
  PrimaryKey,
  Default,
  HasMany,
} from 'sequelize-typescript';
import UserModel from './user.model';

@Table({ tableName: 'Cities', timestamps: true })
class CityModel extends Model<CityModel> {
  @PrimaryKey
  @Default(DataType.UUIDV4)
  @Column(DataType.UUIDV4)
  id!: string;

  @Column(DataType.STRING)
  name!: string;

  @Column(DataType.STRING)
  initials!: string;

  @Column(DataType.STRING)
  status!: string;

  @HasMany(() => UserModel, 'cityId')
  users!: UserModel[];
}

export default CityModel;

```

**Comment by wdzn** [comment]
Same problem. It seems like I need to call `getDataValue`.

**Comment by jrm2k6** [comment]
I created a quick repo here showing the issue: https://github.com/jrm2k6/laughing-octo-enigma

When running the command to execute a findAll this is what I get back:

```
Executing (default): SELECT "uuid", "name", "createdAt", "updatedAt", "deletedAt" FROM "users" AS "users" WHERE ("users"."deletedAt" IS NULL);
[
  users {
    dataValues: {
      uuid: '36497139-77d2-4edb-ba49-481a589f2f8c',
      name: 'John Smith',
      createdAt: 2021-02-22T01:15:35.993Z,
      updatedAt: null,
      deletedAt: null
    },
    _previousDataValues: {
      uuid: '36497139-77d2-4edb-ba49-481a589f2f8c',
      name: 'John Smith',
      createdAt: 2021-02-22T01:15:35.993Z,
      updatedAt: null,
      deletedAt: null
    },
    _changed: {},
    _modelOptions: {
      timestamps: true,
      validate: {},
      freezeTableName: false,
      underscored: false,
      paranoid: true,
      rejectOnEmpty: false,
      whereCollection: null,
      schema: null,
      schemaDelimiter: '',
      defaultScope: {},
      scopes: {},
      indexes: [],
      name: [Object],
      omitNull: false,
      sequelize: [Sequelize],
      hooks: {}
    },
    _options: {
      isNewRecord: false,
      _schema: null,
      _schemaDelimiter: '',
      raw: true,
  1 node_modules/
      attributes: [Array]
    },
    isNewRecord: false,
    uuid: undefined,
    name: undefined
  }
]
```

I also use sequelize-typescript with nestjs and I am not having the same issue but haven't found out if it does something different.

Hope that helps so that we can figure out if we are doing something wrong or not.

Thanks!

**Comment by gideaoms** [comment]
@jrm2k6 The problem is I can not get the value just using `user.name`, but if I use `user.getDataValue('name')` works fine.

I tested it with your github project and the result was the same.

**Comment by jrm2k6** [comment]
@gideaoms Yes, this is why I added a way for anybody to get the same results as we have and see if it is something wrong with our setup or not. Makes it easier for a maintainer to understand the issue and be able to debug if needed.

**Comment by jrm2k6** [comment]
@RobinBuschmann Sorry to tag directly in here but any idea if this is something related to the setup or anything else?

**Comment by jrm2k6** [comment]
Ok I did some more investigating and I think in my example, it boils down to using babel-node vs ts-node.
In here: https://github.com/jrm2k6/laughing-octo-enigma/commit/e745244c1b995f174adf2d48263b30ac52c49e3c
I changed my tsconfig to use commonjs instead of esnext, and use ts-node instead of babel-node.

When I run `node -r ts-node/register index.ts` it lets me access the model properties directly. I am not 100% sure yet but I think with the way my babel-node config is set up, it will generate model properties automatically initialized with undefined, overriding what sequelize-typescript might be doing when it comes to accessing model properties. Still a guess though.

**Comment by gideaoms** [comment]
Yep, it makes sense.

**Comment by triesmon** [comment]
Encountered this issue when upgrading typescript from 3 to 4. Data was only in dataValues but not in the class fields. Setting useDefineForClassFields to true in my tsconfig seems to have fixed it.

**Comment by nathanlepori** [comment]
That's weird. For me doing the exact opposite (setting useDefineForClassFields to false) fixed it. I'm targeting ESNext on Node 14.

**Comment by triesmon** [comment]
I had actually set it to false but just wrote the wrong value in the comment. Thanks.

**Comment by devuxer** [comment]
For me, it had nothing to do with babel (which I wasn't using to begin with).

The solution was entirely to add `"useDefineForClassFields": "false"` to my tsconfig.json.

Thanks, @triesmon!

**Comment by tommoor** [comment]
This appears to be the change in babel that breaks this behavior:

https://github.com/babel/babel/pull/11747

**Comment by ephys** [comment]
Take a look at the updated sequelize documentation regarding the issue with class fields: https://sequelize.org/v6/manual/model-basics.html#caveat-with-public-class-fields

**Comment by tommoor** [comment]
The new documentation is a big improvement over where we were at a week ago, unfortunately the `declare` keyword trick only works with `tsc` tooling and not with babel due to the above linked PR.

**Comment by ephys** [comment]
The fact that babel still emits public class fields tagged with `declare` when using decorators is problematic. Makes https://github.com/sequelize/meetings/issues/16 all the more pressing.

**Comment by tommoor** [comment]
I have worked around this in my project with a model class decorator – it's very ugly, but could provide the basis for another way to fix the issue more formally in sequelize by applying the getter/setters _after_ the model constructor.

https://github.com/outline/outline/blob/main/server/models/decorators/Fix.ts

**Comment by d00rsfan** [comment]
First of all great project! Strange that README.md not updated with 'declare' workaround, as well as with ReturnType<() => Model>; for associated fields.
To prepare PR with "declare" column attributes and "ReturnType" associated attributes?

**Comment by kbirger** [comment]
I landed here with a similar issue. The issue arose when I updated my tsconfig to the [recommended settings for node 20](https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping)

It seems that setting `"lib": ["ES2023"]` causes this issue. Upon some investigation, it actually seems like setting `lib` to this value may switch on `useDefineForClassFields`. After I forced `useDefineForClassFields` off, the issue went away. I would not consider this a resolution though, because it seems like the wrong decision long term. 
TSConfig:
```json
{
  "compilerOptions": {
    "allowJs": false,
    "baseUrl": "./",
    "declaration": true,
    "useDefineForClassFields": false,
    "emitDecoratorMetadata": true,
    // "esModuleInterop": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "incremental": true,
    "module": "CommonJS",
    "moduleResolution": "Node",
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "outDir": "./dist",
    "removeComments": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "ES2017",
    "lib": [
      "ES2023"
    ],
    "typeRoots": [
      "node_modules/@types",
      "src/types"
    ],
    "resolveJsonModule": true,
    "sourceMap": true,
    "inlineSources": true,
    "sourceRoot": "/",
    "allowSyntheticDefaultImports": true
  },
  "exclude": [
    "node_modules",
    "dist",
    "knex/templates",
  ]
}
```

Node: 20.11.0
TypeScript: 5.4.5 (although issue appeared on older version too)
Sequelize: 6.37.3
Sequelize-TypeScript: 2.1.6

**Comment by mopcweb** [comment]
> For me, it had nothing to do with babel (which I wasn't using to begin with).
> 
> The solution was entirely to add `"useDefineForClassFields": "false"` to my tsconfig.json.
> 
> Thanks, @triesmon!

Thanks so much! Spent a lot of time trying to find out a problem, this solved my issue!

**Comment by intellix** [comment]
we just ran into this issue after replacing webpack with rspack and I didn't even notice it was a problem until I ran a query and everything was undefined. We've got around 200 models and thousands of fields and if we try to make the "declare" change, there's going to be at least 1 field that makes it's way to production either in the initial push for it, or someone's rebase.

Unless there's an eslint rule that can prevent it from reaching production I think it's quite dangerous that it's only a warning at runtime and for it to essentially silently fail. Nobody is going to notice an undefined property for days/weeks/months after hours of debugging

I'd love to go with this but it's too dangerous to try. Has anyone found an eslint rule that can enforce it on all properties that extend Model perhaps?  TypeScript only prevents it on the reserved fields: id, createdAt, updatedAt etc, but the custom fields are only added magically by sequelize.
